src/code/code_800EC960.c

```diff
@@ -1203,916 +1233,1078 @@ u8 D_8016B9F4;
+ *      - playing the ocarina memory game
+ * Type 2) Check: tests in 8-note chunks (ocarinaFlag & 0xC000 == 0x0000) eg:
+ *      - validating scarecrow spawn song as adult
+ *      - ocarina prompt for zelda's lullaby, saria's song, Storms, Song of Time, etc...
+ */
+void AudioOcarina_CheckSongsWithMusicStaff(void) {
+    u16 curOcarinaSongFlag;
     u16 pad;
-    u8 inputChanged = 0;
+    u8 noNewValidInput = false;
     u16 pad2;
-    s8 sp57 = 0;
-    u8 i;
+    s8 staffOcarinaPlayingPosOffset = 0;
+    u8 songIndex;
     OcarinaNote* prevNote;
     OcarinaNote* note;
 
-    func_800ECDBC();
+    AudioOcarina_CheckIfStartedSong();
 
-    if (sOcarinaHasStartedSong) {
-        if ((D_80130F2C < 0 ? -D_80130F2C : D_80130F2C) >= 0x15) {
-            D_80130F3C = 0;
-            return;
-        }
+    if (!sOcarinaHasStartedSong) {
+        return;
+    }
 
-        // clang-format off
-        if (sPrevOcarinaNoteVal == sCurOcarinaBtnVal || sCurOcarinaBtnVal == 0xFF) { inputChanged = 1; }
-        // clang-format on
-
-        for (i = sOcarinaSongNoteStartIdx; i < sOcarinaSongCnt; i++) {
-            sh = 1 << i;
-            if (sOcarinaAvailSongs & sh) {
-                D_8016BA50[i] = D_8016BA70[i] + 0x12;
-                if (inputChanged) {
-                    // (pointless if check, this is always true)
-                    if ((D_8016BA50[i] >= D_8016BA70[i] - 0x12) && (D_8016BA50[i] >= D_8016BA70[i] + 0x12) &&
-                        (sOcarinaSongs[i][sLearnSongPos[i]].unk_02 == 0) &&
-                        (sLearnSongLastBtn == sLearnSongExpectedNote[i])) {
-                        D_80131878 = i + 1;
-                        sOcarinaInpEnabled = 0;
-                        D_80130F3C = 0;
-                    }
-                } else if (D_8016BA50[i] >= (D_8016BA70[i] - 0x12)) {
-                    if (sLearnSongLastBtn != 0xFF) {
-                        if (sLearnSongLastBtn == sLearnSongExpectedNote[i]) {
-                            if (i == 12) {
-                                D_8016BA50[i] = 0;
-                            }
-                        } else {
-                            sOcarinaAvailSongs ^= sh;
+    if (ABS_ALT(sCurOcarinaBendIndex) > 20) {
+        sOcarinaFlags = 0;
+        return;
+    }
+
+    // clang-format off
+    if (sPrevOcarinaPitch == sCurOcarinaPitch || sCurOcarinaPitch == OCARINA_PITCH_NONE) { noNewValidInput = true; }
+    // clang-format on
+
+    for (songIndex = sFirstOcarinaSongIndex; songIndex < sLastOcarinaSongIndex; songIndex++) {
+        curOcarinaSongFlag = 1 << songIndex;
+
+        if (sAvailOcarinaSongFlags & curOcarinaSongFlag) {
+            sMusicStaffCurHeldLength[songIndex] = sMusicStaffExpectedLength[songIndex] + 18;
+
+            if (noNewValidInput) {
+                if ((sMusicStaffCurHeldLength[songIndex] >= sMusicStaffExpectedLength[songIndex] - 18) &&
+                    (sMusicStaffCurHeldLength[songIndex] >= sMusicStaffExpectedLength[songIndex] + 18) &&
```

Comment:
I don't quite understand where this number `18` comes from, but are you referring to the ordering here? i.e. expected length on the left and curHeldLength on the right? Or are you just questioning dev logic?

Reply:
The dev logic.
In my understanding just from reading what this PR documents, at some point they planned to also check that the player played the song at the right "pace" and afaict this 18 corresponds to about a second of plus/minus leeway compared to the "perfect" note length
So I would think this wants to check the range `sMusicStaffCurHeldLength[songIndex]` is in like `sMusicStaffExpectedLength[songIndex] - 18 <= sMusicStaffCurHeldLength[songIndex] <= sMusicStaffExpectedLength[songIndex] + 18` but the second check is messed up

and it doesn't matter anyway since `sMusicStaffCurHeldLength[songIndex]` is set to a correct value right before this which makes me say they scrapped that idea